feat: Add package registry to eck

tehbooom opened the pull request:
Resolves #8925 

Elastic Package Registry (EPR) has been highly requested to be added to ECK.

EPR does not have any references since it does not require a license nor any other application.

The following was implemented for EPR 
- defaults to TLS
- Sets the default container image to docker.elastic.co/package-registry/distribution
- Users can set their own images
- Users can update the config following the [reference](https://github.com/elastic/package-registry/blob/main/config.reference.yml)
- Kibana can reference the EPR like Elasticsearch and Enterprise Search
- If Kibana references EPR and TLS is enabled it will populate `xpack.fleet.registryUrl` and set the environment variable `NODE_EXTRA_CA_CERTS` to the path of EPR's CA which is mounted
- If a user provides their own `NODE_EXTRA_CA_CERTS` with a mount the controller will combine the certs appending the EPR's CA to the users specified CA

This was tested with and without setting `NODE_EXTRA_CA_CERTS` using the below manifest
```yaml
apiVersion: epr.k8s.elastic.co/v1alpha1
kind: ElasticPackageRegistry
metadata:
  name: registry
spec:
  version: 9.1.2
  count: 1
  podTemplate:
    spec:
      containers:
      - name: package-registry
        image: docker.elastic.co/package-registry/distribution:lite-9.1.2
---
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: elasticsearch
spec:
  version: 9.1.2
  nodeSets:
  - name: default
    count: 1
---
apiVersion: kibana.k8s.elastic.co/v1
kind: Kibana
metadata:
  name: kibana
spec:
  version: 9.1.2
  count: 1
  elasticsearchRef:
    name: elasticsearch
  packageRegistryRef:
    name: registry
  config:
    telemetry.optIn: false
    xpack.fleet.isAirGapped: true
    xpack.fleet.agents.elasticsearch.hosts: ["https://elasticsearch-es-http.default.svc:9200"]
    xpack.fleet.agents.fleet_server.hosts: ["https://fleet-server-agent-http.default.svc:8220"]
    xpack.fleet.packages:
      - name: system
        version: latest
      - name: elastic_agent
        version: latest
      - name: fleet_server
        version: latest
    xpack.fleet.agentPolicies:
      - name: Fleet Server on ECK policy
        id: eck-fleet-server
        namespace: default
        monitoring_enabled:
          - logs
          - metrics
        unenroll_timeout: 900
        package_policies:
        - name: fleet_server-1
          id: fleet_server-1
          package:
            name: fleet_server
  podTemplate:
    spec:
      containers:
      - name: kibana
        env:
        - name: NODE_EXTRA_CA_CERTS
          value: /custom/user/ca-bundle.crt
        volumeMounts:
        - name: custom-ca
          mountPath: /custom/user
          readOnly: true
      volumes:
      - name: custom-ca
        secret:
          secretName: user-custom-ca-secret
---
apiVersion: v1
kind: Secret
metadata:
  name: user-custom-ca-secret
  namespace: default
type: Opaque
data:
  ca-bundle.crt: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUZtVENDQTRHZ0F3SUJBZ0lVYjVrK2d6V3A5YjljWTV4bkhUcWZNdHFHUXIwd0RRWUpLb1pJaHZjTkFRRUwKQlFBd1hERUxNQWtHQTFVRUJoTUNXRmd4RlRBVEJnTlZCQWNNREVSbFptRjFiSFFnUTJsMGVURWNNQm9HQTFVRQpDZ3dUUkdWbVlYVnNkQ0JEYjIxd1lXNTVJRXgwWkRFWU1CWUdBMVVFQXd3UGRHVnpkQzVsYkdGemRHbGpMbU52Ck1CNFhEVEkxTURneU1ERTRNakl3T0ZvWERUTTFNRGd4T0RFNE1qSXdPRm93WERFTE1Ba0dBMVVFQmhNQ1dGZ3gKRlRBVEJnTlZCQWNNREVSbFptRjFiSFFnUTJsMGVURWNNQm9HQTFVRUNnd1RSR1ZtWVhWc2RDQkRiMjF3WVc1NQpJRXgwWkRFWU1CWUdBMVVFQXd3UGRHVnpkQzVsYkdGemRHbGpMbU52TUlJQ0lqQU5CZ2txaGtpRzl3MEJBUUVGCkFBT0NBZzhBTUlJQ0NnS0NBZ0VBMHljTGVySWR3LzdpbGlKMzVBUEZ4bUx6TFRnNWRhUStWSUttS2lNbStlTTYKanJOY3lnbGphNVFEbHYvMStGUm5hamhrRTBobHoycXEzTjk0U1pYN3M2eHBnQUVzMGVQQ3VaZVBNU2VUYlYyRgp0YlIxNnFuM0JjenVxN3laOXZwdHR3MmJRdkJkY3JzZFU4T2RYUWhGNFd4QUFwODRKYWlMNmkzMlA2K2VPODBwCmh3Z1kwS0F1bzZoZC8zaFpNME14M2MwRmJmU0JHaTUyOHZKODYzUDRXZlEwMWdtUUxVbGl0UlhhTUhiaDRXSm0KOU45c0psUXpnbkNuQjZ6YkZjZ2gweWxrakd0UzBIZEo3eSs3dmE0Q1BqdkxlWGpwTnZuQzRjTmlocnp4Wmw5bQphM0ZVdVpiU0lRekE2ZFlkdkdrT2V3OTJEek1BaTdldU14UDdyYVhRejZmc1N6U1V4N1RjQWl5M2E5VU9Fdi9rCk5NV3VTbDlUMHRRSkhJSzJMc0t0MlVKWVVHWk4wOWU2SUVSTlJOL0FIUjVDbTlhcVQ1Q2ZyQW9JVVhNdUg2S1oKN1JCZFFockRxL2xEQk54bWs5dW44V2lic0NSVnkvVXRJQ3lOSytxbGpGUWZEd01hNkRkd3BjcnpnTWZnU3RTawpLek1LRUJla2N0Q0Q4dHNmTjZYem5USmNBYUJETzFlQWZyT0Z2NG1PTXJqVG90OEYvK3pxN0dXNTlqWTRvdFhMCkY3TnpadFl0eWsvbDRvb2hUZUFuM1ptd1BDMGJFQ1FkTmpTVkZ6ZXJCamE4ZjhacGpKRzNjUllyVmh6YUNsRWMKRU5wbFRHcldVaUVwRDdnTnNlNWNDSnZpQU12NHdwait2QTVVNlA3Z0MxUUtKV2hWS3BVYWcvTmtTSUFCRmtrQwpBd0VBQWFOVE1GRXdIUVlEVlIwT0JCWUVGTWdldEVJajZtRWdsZURGNkVNdUY4NXVnYzdZTUI4R0ExVWRJd1FZCk1CYUFGTWdldEVJajZtRWdsZURGNkVNdUY4NXVnYzdZTUE4R0ExVWRFd0VCL3dRRk1BTUJBZjh3RFFZSktvWkkKaHZjTkFRRUxCUUFEZ2dJQkFEOFU3dm1yWmhHTUZiV2YzRDZlNy84TUwzWEhLRk5TNy9UeWF3U2tvdGVSTVdFbgp1RWhQK2dmbkdUT2ZITFlQeHl5eEJ4U041T29sZHRJclo5dnhBc2dlYWJzSkJaenhQVHpxU09VN3h3b09LcTlRCmdKRUYxL0ZmemFlR1V5dVE2S1ZaZ0QvZ1JPSW42Ri9OUGlzM1pvbUpPOStuVWdTTnNiUm9RYmdPUGdPV3Q3Z1gKVEhuOHJpdUp2OXRPNFBRN09Sa3pubDJYbERlcE9xNVpwSUtkcVl0Rm5MUjF3SllyREZESmt0Q3h6MzFob0FrZwpSVjlSU1BSMFFxZ1JQeFNpNGpXdkNGUk5XTUFJc0NadGJsWExRRUljWGI1YnlsWXV2a3psTTJ4dHlHK3FaRFhMCnFoZDVNeFZIUkpqTzE1VEdpZXFRcUpMVkZyVElhTHFoaXZpQ1pUbDJoVkYxVlpPVG05MU5aeE53M25RL3JyeDgKK2VQV2xTWlZKWXc3SDRkWkx5WTFjRUxLT0YrZDJybVNSZ2pWaHZycUZ3R1M3MUQzYkV4Y0dSakNrOHNQWEZyRwpsOFRzY05RMXBPSGVuNlJhOFhVdGtxU1doZllFb3owZjBEem4wYmt4c2VWaCttS1BHV3QxcHdlemVFTFVwaHE3CmwwSVRLeis1b1lqYWVHTDRia25kcWlpemwzWkc2N0lYL3VyR0dQVUxkLzU1NEtRMFFPMS92S3Y2dE1YMWc0dVMKWHdWc0pzQjlrTUIwRFFxbDhRYmg0UEJ2ZW9RRTZvL3BycXRtWjR1RWdDMCt1cm5paDlCY1FweFNKOUljR1kxTQpBQzRBcG5Pem1CYTFhUVBMcDRaRFIxQXpFK1hXWDd2WWNWYUxleUJxRzRja3dwbUtOUnhpcnJjS2NaMkYKLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo=
---
apiVersion: agent.k8s.elastic.co/v1alpha1
kind: Agent
metadata:
  name: fleet-server
spec:
  version: 9.1.2
  kibanaRef:
    name: kibana
  elasticsearchRefs:
  - name: elasticsearch
  mode: fleet
  fleetServerEnabled: true
  policyID: eck-fleet-server
  deployment:
    replicas: 1
    podTemplate:
      spec:
        serviceAccountName: fleet-server
        automountServiceAccountToken: true
        resources:
          requests:
            cpu: 200m
            memory: 1Gi
          limits:
            cpu: 1
            memory: 2Gi
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: fleet-server
  namespace: default
rules:
- apiGroups: [""]
  resources:
  - pods
  - namespaces
  - nodes
  verbs:
  - get
  - watch
  - list
- apiGroups: ["apps"]
  resources:
    - replicasets
  verbs:
    - get
    - watch
    - list
- apiGroups: ["batch"]
  resources:
    - jobs
  verbs:
    - get
    - watch
    - list
- apiGroups: ["coordination.k8s.io"]
  resources:
  - leases
  verbs:
  - get
  - create
  - update
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: fleet-server
  namespace: default
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: fleet-server
  namespace: default
subjects:
- kind: ServiceAccount
  name: fleet-server
  namespace: default
roleRef:
  kind: ClusterRole
  name: fleet-server
  apiGroup: rbac.authorization.k8s.io
```

tehbooom commented on pkg/controller/kibana/initcontainer/fs_scripts.go:
```diff
@@ -57,6 +57,35 @@ fi
+# CA Certificate Management #
+#############################
+
+# Handle EPR CA certificate consolidation with user-provided NODE_EXTRA_CA_CERTS
```
@jsoriano You could yes use the CA on Kibana that ECK uses however if the user specifies a different CA for EPR then this would fail if we defaulted to Kibana's CA. Using the Kibana's CA would be viable if we are certain that the CA is the same for EPR. Im not sure if there is a good way to figure that out or if the solution here is fine.

barkbay reviewed (commented):
I left a few comments, I'm still testing.

For some reason the container does not seem to listen on `8080` and is killed by the kubelet:

```
  Normal   Pulled     27s (x2 over 2m57s)  kubelet            Container image "docker.elastic.co/package-registry/distribution:9.1.0" already present on machine
  Normal   Created    27s (x2 over 2m57s)  kubelet            Created container: package-registry
  Warning  Unhealthy  27s (x3 over 47s)    kubelet            Startup probe failed: Get "https://10.31.84.6:8080/health": dial tcp 10.31.84.6:8080: connect: connection refused
  Normal   Killing    27s                  kubelet            Container package-registry failed startup probe, will be restarted
```

Nothing obvious in the logs, trying to understand what's wrong in my setup...

barkbay commented:
> For some reason the container does not seem to listen on `8080` and is killed by the kubelet:

I have the same problem when I try to run the e2e test `TestElasticPackageRegistryStandalone`:

```
Containers:
  package-registry:
    Container ID:   containerd://964dc18c8a1b461f7bba941dfc688af4375ce6e00a247f3326cd614622b74b89
    Image:          docker.elastic.co/package-registry/distribution:9.0.5
    Image ID:       docker.elastic.co/package-registry/distribution@sha256:15edf005ee2cb3a9611e1dae535c506134de98f6e90eaaa8419a3161b4c7b858
    Port:           8080/TCP
    Host Port:      0/TCP
    State:          Running
      Started:      Tue, 26 Aug 2025 12:41:15 +0200
    Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Tue, 26 Aug 2025 12:38:55 +0200
      Finished:     Tue, 26 Aug 2025 12:41:15 +0200
    Ready:          False
    Restart Count:  1
[...]
  Normal   Killing    2m3s                  kubelet            Container package-registry failed startup probe, will be restarted
  Warning  Unhealthy  3s (x4 over 2m23s)    kubelet            Startup probe failed: Get "https://10.31.86.16:8080/health": dial tcp 10.31.86.16:8080: connect: connection refused
```

barkbay commented on pkg/controller/kibana/pod.go:
```diff
@@ -249,3 +292,45 @@ func GetKibanaBasePath(kb kbv1.Kibana) (string, error) {
+	eprAssocConf, err := kb.EPRAssociation().AssociationConf()
+	if err != nil {
+		return nil, err
+	}
+
+	if !eprAssocConf.CAIsConfigured() {
+		return builder, nil
+	}
+
+	// Volume mounting is now handled in buildVolumes()
+	// Check if user has provided NODE_EXTRA_CA_CERTS
+	mainContainer := builder.MainContainer()
+	userProvidedNodeExtraCACerts := false
+	if mainContainer != nil {
+		for _, env := range mainContainer.Env {
+			if env.Name == nodeExtraCACerts {
+				userProvidedNodeExtraCACerts = true
+				break
+			}
+		}
+	}
+
+	if !userProvidedNodeExtraCACerts {
+		// User hasn't provided NODE_EXTRA_CA_CERTS, set it to EPR CA path
+		builder, _ = builder.WithNewEnv(corev1.EnvVar{Name: nodeExtraCACerts, Value: eprCertsVolumeMountPath + "/ca.crt"})
+	} else {
+		// User provided NODE_EXTRA_CA_CERTS, init container will create combined bundle
+		// Override the user-provided value by directly modifying the container
+		mainContainer := builder.MainContainer()
+		if mainContainer != nil {
+			for i, env := range mainContainer.Env {
+				if env.Name == nodeExtraCACerts {
+					mainContainer.Env[i].Value = "/usr/share/kibana/config/combined-ca-bundle.crt"
+					break
+				}
+			}
+		}
+	}
+	return builder, nil
+}
```
Added a godoc and tried to reduce some nested `if` `then` `else`:

 ```suggestion
// withEPRCertsVolume sets NODE_EXTRA_CA_CERTS environment variable to point to either ca.crt or combined-ca-bundle.crt if the user has already set NODE_EXTRA_CA_CERTS in the podTemplate.
func withEPRCertsVolume(builder *defaults.PodTemplateBuilder, kb kbv1.Kibana, userProvidedNodeExtraCACerts bool) (*defaults.PodTemplateBuilder, error) {
	eprAssocConf, err := kb.EPRAssociation().AssociationConf()
	if err != nil {
		return nil, err
	}

	if !eprAssocConf.CAIsConfigured() {
		return builder, nil
	}

	if !userProvidedNodeExtraCACerts {
		// User hasn't provided NODE_EXTRA_CA_CERTS, set it to EPR CA path
		return builder.WithEnv(corev1.EnvVar{Name: nodeExtraCACerts, Value: eprCertsVolumeMountPath + "/ca.crt"}), nil
	}

	// User provided NODE_EXTRA_CA_CERTS, init container will create combined bundle
	// Override the user-provided value by directly modifying the container
	mainContainer := builder.MainContainer()
	if mainContainer == nil {
		return builder, nil
	}

	for i, env := range mainContainer.Env {
		if env.Name == nodeExtraCACerts {
			mainContainer.Env[i].Value = "/usr/share/kibana/config/combined-ca-bundle.crt"
			break
		}
	}

	return builder, nil
}
```

(See my 2 next comments)

barkbay commented on pkg/controller/kibana/pod.go:
```diff
@@ -173,12 +174,54 @@ func NewPodTemplateSpec(
 	}
 
-	return builder.WithInitContainerDefaults().PodTemplate, nil
+	// Pass original user-provided NODE_EXTRA_CA_CERTS to init container if EPR association exists
+	var additionalInitEnvVars []corev1.EnvVar
+	if kb.Spec.PackageRegistryRef.IsDefined() {
+		userNodeExtraCACerts := getUserNodeExtraCACerts(kb)
+		if userNodeExtraCACerts != "" {
+			// Pass original user path to init container so it can copy from there
+			additionalInitEnvVars = append(additionalInitEnvVars, corev1.EnvVar{
+				Name:  nodeExtraCACerts,
+				Value: userNodeExtraCACerts,
+			})
+		}
+	}
+
+	return builder.WithInitContainerDefaults(additionalInitEnvVars...).PodTemplate, nil
```
```suggestion
	return builder.WithInitContainerDefaults(additionalInitEnvVars...).PodTemplate, nil
```

barkbay commented on pkg/controller/kibana/pod.go:
```diff
@@ -173,12 +174,54 @@ func NewPodTemplateSpec(
 			WithInitContainers(keystore.InitContainer)
 	}
 
+	if kb.Spec.PackageRegistryRef.IsDefined() {
+		builder, err = withEPRCertsVolume(builder, kb)
+		if err != nil {
+			return corev1.PodTemplateSpec{}, err
+		}
+	}
+
```
```suggestion
	var additionalInitEnvVars []corev1.EnvVar
	if kb.Spec.PackageRegistryRef.IsDefined() {
		userNodeExtraCACerts := getUserNodeExtraCACerts(kb)
		builder, err = withEPRCertsVolume(
			builder,
			kb,
			userNodeExtraCACerts != "",
		)
		if err != nil {
			return corev1.PodTemplateSpec{}, err
		}
		if userNodeExtraCACerts != "" {
			// Pass original user path to init container so it can copy from there
			additionalInitEnvVars = append(additionalInitEnvVars, corev1.EnvVar{
				Name:  nodeExtraCACerts,
				Value: userNodeExtraCACerts,
			})
		}
	}
```

barkbay commented on pkg/controller/kibana/pod.go:
```diff
@@ -173,12 +174,54 @@ func NewPodTemplateSpec(
+				return env.Value
+			}
+		}
+		break
```
```suggestion
```

barkbay commented on pkg/controller/packageregistry/pod.go:
```diff
@@ -0,0 +1,140 @@
+	}
+}
+
+// readinessProbe is the readiness probe for the maps container
```
```suggestion
// readinessProbe is the readiness probe for the epr container
```

barkbay commented on pkg/controller/packageregistry/pod.go:
```diff
@@ -0,0 +1,140 @@
+const (
+	HTTPPort                 = 8080
+	configHashAnnotationName = "packageregistry.k8s.elastic.co/config-hash"
+	EPRTLSKey                = "EPR_TLS_KEY"
+	EPRTLSCert               = "EPR_TLS_CERT"
+	EPRAddress               = "EPR_ADDRESS"
```
nit: we are already in a package named `packageregistry`, we can remove the `EPR` prefix:
```suggestion
	TLSKeyEnvName            = "EPR_TLS_KEY"
	TLSCertEnvName           = "EPR_TLS_CERT"
	AddressEnvName           = "EPR_ADDRESS"
```

barkbay commented on pkg/controller/packageregistry/pod.go:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	eprVars := []corev1.EnvVar{
+		corev1.EnvVar{Name: EPRAddress, Value: "0.0.0.0:8080"},
```
```suggestion
		{Name: EPRAddress, Value: "0.0.0.0:8080"},
```

tehbooom commented on pkg/controller/common/version/version.go:
```diff
@@ -34,9 +34,10 @@ var (
-	SupportedLogstashVersions       = MinMaxVersion{Min: From(8, 12, 0), Max: From(9, 99, 99)}
+	SupportedFleetModeAgentVersions  = MinMaxVersion{Min: MustParse("7.14.0-SNAPSHOT"), Max: From(9, 99, 99)}
+	SupportedMapsVersions            = MinMaxVersion{Min: From(7, 11, 0), Max: From(9, 99, 99)}
+	SupportedPackageRegistryVersions = MinMaxVersion{Min: From(7, 15, 1), Max: From(9, 99, 99)}
```
@jsoriano I dont think we have ubi images for EPR correct?

jsoriano replied:
Correct, we don't have. There is an open issue about this https://github.com/elastic/package-registry/issues/1085.

barkbay replied:
Any chance to get this for the next release? I think we expect to have UBI images of our product, IIRC all images pulled by ECK on OpenShift must be based on UBI.

tehbooom commented on config/recipes/packageregistry/03-ingress.yaml:
```diff
@@ -0,0 +1,32 @@
+apiVersion: networking.k8s.io/v1
+kind: Ingress
```
I was thinking the same. I kept ingress because I didnt think it could hurt however, every time I deploy EPR i never directly access it only Kibana ever accesses it. Thus I think it is fair to remove Ingress.

tehbooom commented:
@barkbay With the startup taking anywhere from 2-10+ minutes do we want to  set the initial delay to still be 120 seconds but increase the failure threshold, and timeoutseconds to account for the 10+ minutes?

Something like this where the max amount of time is 10 minutes (600 seconds)
```golang
// startupProbe is the startup probe for the packageregistry container
func startupProbe(useTLS bool) corev1.Probe {
	scheme := corev1.URISchemeHTTP
	if useTLS {
		scheme = corev1.URISchemeHTTPS
	}
	return corev1.Probe{
		FailureThreshold:    16,
		InitialDelaySeconds: 120,
		PeriodSeconds:       10,
		SuccessThreshold:    1,
		TimeoutSeconds:      30,
		ProbeHandler: corev1.ProbeHandler{
			HTTPGet: &corev1.HTTPGetAction{
				Port:   intstr.FromInt(HTTPPort),
				Path:   "/health",
				Scheme: scheme,
			},
		},
	}
}
```

barkbay commented:
> It 100% is the startup probe time. I had a lot of issues with this and at one point had it set at 10 minutes. I think 7 minutes might be a good in between?

I think my question would be "why do we need a startup probe"? Maybe a readiness probe is enough?

tehbooom commented:
> > It 100% is the startup probe time. I had a lot of issues with this and at one point had it set at 10 minutes. I think 7 minutes might be a good in between?
> 
> I think my question would be "why do we need a startup probe"? Maybe a readiness probe is enough?

Ahh yes I guess the application has already started therefore a startup probe doesnt make sense here.

tehbooom commented on pkg/controller/packageregistry/config.go:
```diff
@@ -0,0 +1,96 @@
+
+func defaultConfig() *settings.CanonicalConfig {
+	return settings.MustCanonicalConfig(map[string]interface{}{
+		"package_paths": []string{"/packages/package-registry", "/packages/package-storage"},
```
@jsoriano tested and validated [here ](https://github.com/elastic/cloud-on-k8s/pull/8800/files#diff-8f5ff9ac996a46622dae69a2366a344246871d0e28bac591e840185b2baa883d).

Created a configmap with my netbox integration (still in review) was able to add that to an agent.

tehbooom commented:
> @tehbooom 👋 please let me know when you need another review, thanks!

@barkbay  

Added EPR to ECK diagnostics [here](https://github.com/elastic/eck-diagnostics/pull/346).

I still need to update our documentation. I know we changed how we do documentation so if you could please point me in the right direction where ECK docs live that would be great. This PR is ready for another review. Thanks!

barkbay commented on Makefile:
```diff
@@ -422,7 +422,7 @@ E2E_STACK_VERSION          ?= 9.0.0
 export TESTS_MATCH         ?= "^Test"
 export E2E_JSON            ?= false
-TEST_TIMEOUT               ?= 15m
+TEST_TIMEOUT               ?= 20m
```
Moving the discussion here: https://github.com/elastic/cloud-on-k8s/pull/8800#discussion_r2301278209

pebrc replied:
Can we undo this change now that we use a smaller image?

barkbay commented:
E2E tests are still failing with:

```
  Warning  Failed               12m   kubelet            Failed to pull image "docker.elastic.co/package-registry/distribution:9.1.2": failed to pull and unpack image "docker.elastic.co/package-registry/distribution:9.1.2": failed to extract layer sha256:57bbe197467b8b19ba0705f05ee41860ff3bb44020ed5986df96bfed4614e630: write /var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/396/fs/packages/package-storage/security_detection_engine-8.6.8.zip: no space left on device
```

We also have to find a solution for this. IIUC https://github.com/elastic/package-registry/pull/1335 would be the way to go? I'll also try to check if we can increase the disk on GKE nodes, but it would still mean that we may have to skip other providers (AWS, Azure, Kind...).

pebrc reviewed (commented):
Approach looks reasonable with one exception: I am also not a big fan of of the NODE_EXTRA_CA_CERTS init container hackery but don't see a good alternative right now if we want to be backwards compatible. 

Main blocker to merge this imo is the lack of UBI images for the package registry.

I could not get the recommended default image to work, as the k8s nodes in my environment ran out of ephemeral storage while trying to pull the image. I wonder how viable this image is for any environment? I was successful with the `lite` image though.

tehbooom commented on pkg/controller/kibana/initcontainer/fs_scripts.go:
```diff
@@ -57,6 +57,35 @@ fi
+if [[ -f "` + volume.EPRCACertPath + `" ]]; then
+	echo "EPR CA certificate found, checking for user-provided CA bundle..."
+	
+	# Check if user provided their own NODE_EXTRA_CA_CERTS (from init container env)
```
> I think one issue with that would be that there might be already users that we would break if we stopped supporting user defined `NODE_EXTRA_CA_CERTS`

Agreed. However, if we did use `KIBANA_EXTRA_CA_CERTS` what would it make it easier to debug and implement? You are still checking if the user provided extra CAs.

tehbooom reviewed (commented):
Ive update the test image to be the `lite` image since others have stated it works with that image

jsoriano commented:
> > Main blocker to merge this imo is the lack of UBI images for the package registry.
> 
> This blocker has been addressed in [elastic/package-registry#1451](https://github.com/elastic/package-registry/pull/1451), which is now merged.

In this PR we are using the Package Registry distribution images. To support UBI there we would also need to update https://github.com/elastic/package-storage-infra/blob/13bf4e9ba03c028b16ed37772cd0d1afaa45af4f/.buildkite/scripts/build_distributions.sh.

naemono commented on config/recipes/packageregistry/epr-eck.yaml:
```diff
@@ -0,0 +1,134 @@
+  name: registry
+spec:
+  version: 9.1.2
+  count: 1
```
Should we note here links to the `lite` container instead?

naemono commented on pkg/controller/packageregistry/config.go:
```diff
@@ -0,0 +1,96 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:   epr.Namespace,
+			Name:        ConfigName(epr.Name),
+			Labels:      commonlabels.AddCredentialsLabel(maps.Clone(meta.Labels)),
```
This does not contain credentials, and can be removed: 

```
❯ kc view-secret -n elastic-epr registry-epr-config
Viewing only available key: config.yml
cache_time:
    catch_all: 10m
    categories: 10m
    index: 10s
    search: 10m
package_paths:
    - /packages/package-registry
    - /packages/package-storage
```

naemono commented on pkg/controller/packageregistry/config.go:
```diff
@@ -0,0 +1,96 @@
+			"index":      "10s",
+			"search":     "10m",
+			"categories": "10m",
+			"catch_all":  "10m",
```
Are these always constant, and never need to be "tweaked"?

naemono commented on pkg/controller/packageregistry/controller.go:
```diff
@@ -0,0 +1,379 @@
+func newReconciler(mgr manager.Manager, params operator.Parameters) *ReconcilePackageRegistry {
+	client := mgr.GetClient()
+	return &ReconcilePackageRegistry{
+		Client:         client,
```
```suggestion
	return &ReconcilePackageRegistry{
		Client:         mgr.GetClient(),
```

naemono reviewed (commented):
My understanding is that we will take ownership of this PR and move it towards a merging state. Review notes are primarily for our own purposes.

naemono commented on pkg/apis/epr/v1alpha1/webhook.go:
```diff
@@ -0,0 +1,91 @@
+	}
+
+	if len(errors) > 0 {
+		validationLog.V(1).Info("failed validation", "errors", errors)
```
Fyi, I removed this.

naemono commented:
> > > Main blocker to merge this imo is the lack of UBI images for the package registry.
> > 
> > 
> > This blocker has been addressed in [elastic/package-registry#1451](https://github.com/elastic/package-registry/pull/1451), which is now merged.
> 
> In this PR we are using the Package Registry distribution images. To support UBI there we would also need to update https://github.com/elastic/package-storage-infra/blob/13bf4e9ba03c028b16ed37772cd0d1afaa45af4f/.buildkite/scripts/build_distributions.sh.

The beginnings of this needed PR are [here](https://github.com/elastic/package-storage-infra/pull/1072) @jsoriano

naemono commented:
buildkite test this -f p=gke,E2E_TAGS=epr

naemono commented:
> And more fun on ocp:

And we don't set runasuser/runasgroup for any CRD either:

```
containers[0].runAsUser: Invalid value: 1000: must be in the ranges: [1000730000, 1000739999]
```

I'm fixing all of these issues....

pkoutsovasilis commented on pkg/controller/association/controller/kibana_epr.go:
```diff
@@ -0,0 +1,104 @@
+			return "", false, err
+		}
+		eprVersionResponse := &eprVersionResponse{}
+		ver, isServerless, err := info.Version("/api/epr/v1/internal/version", eprVersionResponse)
```
some extra thoughts on that; AFAICT this path `/api/epr/v1/internal/version` doesn't exist at the elastic package registry service, e.g. [here](https://epr.elastic.co/). It reminds me of a kibana service path but even there I get a 404. Also, do we rely support an external EPR for Kibana? AFAIK EPR doesn't need a username password and `association.GetUnmanagedAssociationConnectionInfoFromSecret` will look for these

pkoutsovasilis commented on pkg/apis/epr/v1alpha1/epr_types.go:
```diff
@@ -0,0 +1,120 @@
+	EPRContainerName = "package-registry"
+	// Kind is inferred from the struct name using reflection in SchemeBuilder.Register()
+	// we duplicate it as a constant here for practical purposes.
+	Kind = "ElasticPackageRegistry"
```
the PR to do so is here https://github.com/elastic/eck-diagnostics/pull/359, so I am resolving this one. please feel free to unresolve if you think there is something else missing

pkoutsovasilis commented:
buildkite test this -f p=gke,E2E_TAGS=epr

pebrc commented on pkg/controller/packageregistry/controller.go:
```diff
@@ -0,0 +1,378 @@
+	if epr.Spec.HTTP.TLS.Enabled() {
+		var tlsCertSecret corev1.Secret
+		tlsSecretKey := types.NamespacedName{Namespace: epr.Namespace, Name: certificates.InternalCertsSecretName(eprv1alpha1.Namer, epr.Name)}
+		if err := c.Get(context.Background(), tlsSecretKey, &tlsCertSecret); err != nil {
```
Must not instaniate new context but re-use existing context: 
- APM tracing 
- logger is in context 
- proper cancellation propagation

pkoutsovasilis replied:
changed here 8f49e4f9246b705c66dd0c90c23311419c31ddd8

pebrc commented on pkg/apis/packageregistry/v1alpha1/labels.go:
```diff
@@ -0,0 +1,17 @@
+func (m *PackageRegistry) GetIdentityLabels() map[string]string {
+	return map[string]string{
+		commonv1.TypeLabelName:                "package-registry",
+		"packageregistry.k8s.elastic.co/name": m.Name,
```
Move the constants here so we avoid the string literal or move the function.

pebrc commented on cmd/manager/main.go:
```diff
@@ -713,6 +715,11 @@ func startOperator(ctx context.Context) error {
 		return err
 	}
 
+	isOpenshift, err := isOpenShift(clientset)
```
Move this further up above L691 and use it as an input for `determineSetDefaultSecurityContext` so we don't have to run the detection twice.

pkoutsovasilis replied:
changed here ab208bebc3bfb38a32ed55019eabd3f79c8c3a11

pkoutsovasilis commented:
buildkite test this -f p=gke,E2E_TAGS='epr kb',t='TestKibanaCrossNSEPRAssociation|TestEPRAssociation|TestKibanaAssociationWithNonExistentEPR|TestPackageRegistry*'

pkoutsovasilis commented on pkg/controller/packageregistry/label/label.go:
```diff
@@ -0,0 +1,28 @@
+
+const (
+	// NameLabelName used to represent a EPR in k8s resources
+	NameLabelName = "packageregistry.k8s.elastic.co/name"
```
For any other controller we don't define the labels in the api package, thus I got inspired by the kibana one and did this change 4f4babb64994ce60383f254d0e51c8d5a7fe93d7. Lmk what you think

pkoutsovasilis commented on pkg/controller/packageregistry/pod.go:
```diff
@@ -0,0 +1,130 @@
+			},
+			Privileged:             ptr.To(false),
+			ReadOnlyRootFilesystem: ptr.To(true),
+			RunAsNonRoot:           ptr.To(true),
```
I did experiment with OpenShift today and I validated that if we don't set `RunAsNonRoot=true` it gets automatically set to that. Thus here 156599bc287335eb915a22979496c99fe7f82021 I did remove the previous hacky approach of handling differently the securityContext between k8s and ocp. This means that for k8s, EPR will run as root which can be altered in a follow-up PR for newer versions given that this PR https://github.com/elastic/package-registry/pull/1503 is now merged

naemono commented on pkg/controller/packageregistry/controller_test.go:
```diff
@@ -0,0 +1,302 @@
+	assertObservedGeneration := func(r k8s.Client, expected int) {
+		var ems v1alpha1.PackageRegistry
+		require.NoError(t, r.Get(context.Background(), types.NamespacedName{Name: nsnFixture.Name, Namespace: nsnFixture.Namespace}, &ems))
+		require.Equal(t, int64(expected), ems.Status.ObservedGeneration)
```
```suggestion
		var epr v1alpha1.PackageRegistry
		require.NoError(t, r.Get(context.Background(), types.NamespacedName{Name: nsnFixture.Name, Namespace: nsnFixture.Namespace}, &epr))
		require.Equal(t, int64(expected), epr.Status.ObservedGeneration)
```

pkoutsovasilis replied:
fixed here f3873d4817a16a844605156683006b8cd117c36e

naemono reviewed (commented):
I don't see any major issues with this PR, but I did see 
1. a typo
2. a potential issue that seems similar to something I dealt with in the autoops PR.

pkoutsovasilis commented on pkg/controller/common/container/container.go:
```diff
@@ -123,3 +128,29 @@ func getUBISuffix(ver version.Version) string {
+		return fmt.Sprintf("%s/%s:lite-9.1.8-ubi", containerRegistry, PackageRegistryImage)
+	case v.Major == 9 && v.Minor > 1 && v.LT(version.From(9, 2, 2)):
+		// Fallback to 9.2.2-ubi for 9.2.x versions below 9.2.2
+		return fmt.Sprintf("%s/%s:lite-9.2.2-ubi", containerRegistry, PackageRegistryImage)
```
```
// Since UBI images are only offered from certain versions onwards,
// fallback to tested backwards-compatible versions for unsupported releases
```
isn't that enough of justification [here](https://github.com/tehbooom/cloud-on-k8s/blob/db19f288723dd7bf40d5cb887a823c45362ebbfa/pkg/controller/common/container/container.go#L140-L143) ?
